### PR TITLE
refactor(convert): split extract_block_style into per-property modules

### PR DESCRIPTION
**Pull request by mitsuru**
## 概要

`extract_block_style` (約 245 行のモノリシック関数) を `convert/style/<prop>.rs` 配下のプロパティ別モジュールに分割。CSS プロパティを追加する WPT ブランチが単一関数で衝突する状況を解消。

beads: `fulgur-zcmt` / plan: `docs/plans/2026-04-27-convert-style-phase2.md`

## 最終構造

```text
crates/fulgur/src/convert/style/
  mod.rs           72 行  StyleContext + extract_block_style + absolute_to_rgba
  box_metrics.rs   19 行  border_widths + padding (layout-only)
  border.rs        76 行  border_color + border_radii + border_styles
  background.rs   566 行  bg_color + bg_image_layers + 12 gradient/url helpers
  shadow.rs        39 行  box_shadows
  overflow.rs      53 行  overflow_x/y (overflow::map に unit test 5 件)
  opacity.rs       26 行  extract_opacity_visible (free function)
```

`convert/mod.rs` は 1806 → 1095 行（-711 行）。

## issue 記述からの意図的な逸脱

| 逸脱 | 理由 |
|---|---|
| `absolute_to_rgba` を `style/mod.rs` に残置（`background.rs` ではなく） | shadow / border / background から呼ばれる共有 helper のため |
| `transform.rs` 不採用 | `BlockStyle` に transform field なし。`maybe_wrap_transform` は別経路で `Pageable` を wrap |
| `overflow.rs` を新規追加 | issue 未記載だが `extract_block_style` 内に存在するため |
| `box_metrics::apply_to(&Layout)` のみ非統一 signature | styles を必要としないため、未使用引数を避ける |
| `try_convert_lp_to_bg` も同時移動（spec の 13 helpers → 実 14 helpers） | radial/conic gradient のみから呼ばれる helper、移動しないと dead_code |

## 設計上の決定事項

- **Subagent-driven 実装**: 8 タスク × per-task review で実施。Task 1 は「抽出ゼロの保険コミット」とし、VRT 退行時の切り分けを容易に
- **VRT byte-identical を全 commit で死守**: per-task verification で `FONTCONFIG_FILE` 経由 VRT 実行
- **per-task crate-local clippy gate**: workspace-wide clippy は Task 8 のみだが、`cargo clippy -p fulgur --lib --all-targets -- -D warnings` を per-task で実行（Task 3 review で混入した auto-deref lint を捕捉した実績）
- **`convert/style/` ディレクトリ名維持**: Stylo extern crate `style` と衝突するが、影響範囲は `use super::*;` を使う 4 sibling のみで一回限り。`convert/mod.rs:35-39` に collision 説明コメントを設置
- **`pub(in crate::convert)` で `extract_opacity_visible` を最小権限化**: `convert` 外（`paragraph`, `render` など）からの誤呼び出しを防ぐ

## Out of scope (Phase 3 候補)

- `list_marker` / `pseudo` / abs-position helper module の更なる分割
- `BlockStyle` に transform field を追加して `maybe_wrap_transform` を style pipeline に統合
- `extract_asset_name` を `asset.rs` 寄せ（現状 `convert/mod.rs` に残置、`get_text_color` / `get_text_decoration` がここに居るため）
- `convert.rs` ↔ `blitz_adapter.rs` 境界の縮小（fulgur-x92a で別管理）

## Test plan

- [x] `cargo test -p fulgur` (lib 809 / gcpm integration 25 すべて pass)
- [x] `cargo clippy --workspace --all-targets -- -D warnings` clean
- [x] `cargo fmt --check` clean
- [x] VRT byte-identical (`FONTCONFIG_FILE="\$PWD/examples/.fontconfig/fonts.conf" cargo test -p fulgur-vrt`)
- [x] `cargo build --workspace` clean (全 binding crate も含む)

---

<a href="https://app.devin.ai/review/fulgur-rs/fulgur/pull/252" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **Refactor**
  * Reorganized style conversion for more reliable and maintainable rendering.

* **New Features**
  * Improved background handling: images, gradients, and color layers render more accurately.
  * Better border and box-metrics extraction for precise layout.
  * Enhanced box-shadow, opacity, and overflow interpretation to match computed styles.

* **Tests**
  * Added end-to-end rendering smoke tests covering shadows, gradients, and background-image edge cases.

* **Documentation**
  * Added a Phase 2 implementation plan for the style conversion work.

### CODE REVIEW

File: crates/fulgur/src/convert/style/opacity.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    use style::properties::longhands::visibility::computed_value::T as Visibility;
+    node.primary_styles()
+        .map(|s| {
+            let opacity = s.clone_opacity();
+            let v = s.clone_visibility();
+            let visible = v != Visibility::Hidden && v != Visibility::Collapse;
+            (opacity, visible)
+        })
+        .unwrap_or((1.0, true))
+}
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply crates/fulgur/src/convert/style/opacity.rs:28]: Dismiss します。`opacity.rs` は **lib coverage 100%** (`cargo llvm-cov nextest -p fulgur` で確認済み) を `crates/fulgur/tests/opacity_visibility_test.rs` 経由で達成しており、追加カバレッジゲインはありません。

`extract_opacity_visible` は `&blitz_dom::Node` を受け取りますが、Node fake の構築には Stylo の `ComputedValues` + `AtomicRef` wrapper + Blitz の Document/parser scaffolding が必要で、純粋関数 `overflow::map` のような軽量な unit test に収まりません。プロジェクトとしての規約「純関数 → unit test」は overflow::map / map_border_style / box_metrics::apply_to で守られていますが、`extract_opacity_visible` は内部で Stylo computed-values にアクセスする bridge 関数であり pure helper の範疇から外れます。

3 ケース (`hidden` / `collapse` / `primary_styles() == None`) は既存の end-to-end test で網羅:
- `test_visibility_hidden_*` 系テスト群が hidden→visible=false 経路を叩く
- `visibility:collapse` は `crates/fulgur-vrt/goldens/fulgur/visibility/*.pdf` reftest で網羅
- `primary_styles() == None` 経路は `<html>` の implicit body や text node でフォールバックされる ((1.0, true)) のを `cargo test -p fulgur --lib` の各 render path テストが暗黙に通る (`opacity` が 1.0 のまま skip されないことを確認)

このため追加 unit test の便益はなく、テスト二重化を避けるために dismiss します。

File: docs/plans/2026-04-27-convert-style-phase2.md [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+|---|---|---|
+| Move all 11 bg helpers (incl. `absolute_to_rgba`) to `background.rs` | `absolute_to_rgba` is also used by `shadow` + `border` | Keep `absolute_to_rgba` in `style/mod.rs` as `pub(super) fn`, move only the 10 bg-specific helpers to `background.rs` (also `resolve_conic_gradient` which the issue text missed) |
+| Add `transform.rs` | `BlockStyle` has no transform field. Transform is handled by `maybe_wrap_transform` (`convert/mod.rs:315`) which wraps a `Pageable`, not a `BlockStyle` field | **Drop `transform.rs` from this phase.** `maybe_wrap_transform` stays where it is |
+| Modules: `box_metrics / border / background / shadow / transform / opacity` | `extract_block_style` also fills `overflow_x` / `overflow_y`, which is missing from the issue list | **Add `overflow.rs`.** Same shape as the others |
+| Uniform 5-param `apply_to` signature on every module | `box_metrics` does not need `styles` / `current_color` / `assets`; `overflow` does not need `current_color` / `layout` / `assets` | **Use a small `StyleContext` bundle for the inner-styles modules; `box_metrics` keeps a thin `&Layout`-only signature.** Better than passing unused params |
+
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply docs/plans/2026-04-27-convert-style-phase2.md:27]: e25e170 で修正しました。Deviations 表の該当行を最終的な内訳に書き直しました:

- `absolute_to_rgba` は `style/mod.rs` に `pub(super) fn` として残置 (1 helper stays)
- 以下 12 個の helper を `background.rs` に private `fn` として移動: `resolve_linear_gradient`, `resolve_color_stops`, `resolve_radial_gradient`, `resolve_conic_gradient` (issue 記述漏れ), `map_extent`, `convert_bg_size`, `convert_lp_to_bg`, `try_convert_lp_to_bg` (sole callers `resolve_radial_gradient` / `resolve_conic_gradient` も同時移動), `convert_bg_position`, `convert_bg_repeat`, `convert_bg_origin`, `convert_bg_clip`
- 合計 13 シンボル (= 1 残置 + 12 移動)

`resolve_conic_gradient` の扱いも明示し、issue 記述「11 helpers」と実装「13 symbols」の差分が読み手に追える形にしました。

File: docs/plans/2026-04-27-convert-style-phase2.md [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+Actually — since callers already use `extract_opacity_visible` unqualified, and they live in sibling modules of `style`, the cleanest route is a re-export at `convert/mod.rs`:
+
+```rust
+pub(super) use self::style::opacity::extract_opacity_visible;
+```
+
+That way no caller changes its import.
+
+Verify no caller said `crate::convert::extract_opacity_visible` (which would still resolve via the re-export). `grep -rn "extract_opacity_visible" crates/fulgur/src/`.
+
```
[the remark this answers is not in the document]
Comment (mitsuru) [reply docs/plans/2026-04-27-convert-style-phase2.md]: e25e170 で修正しました。Task 7 step 7.2 を実装に揃えて書き直しました:

- 唯一の re-export 場所は `crates/fulgur/src/convert/style/mod.rs` の `pub(super) use opacity::extract_opacity_visible;`
- `convert/mod.rs:38` の既存 `use self::style::{absolute_to_rgba, extract_block_style, extract_opacity_visible};` がこの再エクスポート経由で `extract_opacity_visible` を解決する
- **`convert/mod.rs` 側に追加の re-export は要らない** ことを明示

これで「どこに re-export を書くべきか」の混乱が消えるはずです。
